### PR TITLE
tech debt: move connection and async client APIs to thread local cluster

### PR DESCRIPTION
**Pull request by mattklein123**
This is part of several follow ups to #13906. This change moves various
functions from the cluster API to the thread local cluster API. This
simplifies error handling and also will increase performance slightly
as many instances of double map lookups are now a single lookup.

Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@alyssawilk assigning you since you have been pretty close to this code. I didn't realize how awful and large this was going to be when I started, but it's the right thing to do and will make a final follow up change that I have simpler. I don't think there is any great way to break this apart but the actual prod changes should be easy to review.

File: include/envoy/upstream/thread_local_cluster.h
Comment (alyssawilk) [inline include/envoy/upstream/thread_local_cluster.h:46]: Mind adding @params and @returns while you're in here?

File: test/common/upstream/cluster_update_tracker_test.cc
Comment (alyssawilk) [inline test/common/upstream/cluster_update_tracker_test.cc:64]: optional: arguably these should be ASSERTs, since the next line will fail if we don't early return.  
Given you're just changing function signatures your call if you want to clean it up.

**Review by alyssawilk — COMMENTED** [review]
Awesome clean-up!  The one change I was having trouble following was why this removes the necessity of the various config validation bits.   Just to confirm, they were always shim classes required by prior API organization but not actually validating any config?

**Comment by mattklein123** [comment]
> The one change I was having trouble following was why this removes the necessity of the various config validation bits. Just to confirm, they were always shim classes required by prior API organization but not actually validating any config?

They were shims previously, but now I'm actually questioning whether I did this correctly, because it's possible that the way I did it will actually make real networking calls when validating config which seems wrong. Let me look at this again and make sure I did this correctly and also fix your other comments.

**Comment by mattklein123** [comment]
@alyssawilk updated. Thanks for poking me on the validation code. It's actually fairly confusing and not well defined/tested what the validation code is supposed to do but I think the way I have it now is more appropriate. I also fixed more cases that should have had error handling but did not.

File: source/extensions/filters/http/squash/squash_filter.cc
Comment (alyssawilk) [inline source/extensions/filters/http/squash/squash_filter.cc:280]: do we need another thread local cluster check here?

File: source/server/config_validation/cluster_manager.h
Comment (alyssawilk) [inline source/server/config_validation/cluster_manager.h:60]: Awesome, this makes more sense to me now, thanks

**Review by alyssawilk — APPROVED** [review]
LGTM modulo that one squash issue  (since I'm out tomorrow)
If you don't need that check, feel free to submit.  If you do, hopefully someone else can stamp that one liner :-)

File: source/extensions/filters/http/squash/squash_filter.cc
Comment (mattklein123) [reply source/extensions/filters/http/squash/squash_filter.cc:280]: Technically I think yes, but I couldn't quickly sort out how to test this and this code is not clearly used/owned so I will admit I punted. Also, it's guarded behind a previous check, so it would be a rare condition to hit this becoming null but it could happen. Let me add a TODO for this to follow up later since it's a pre-existing bug.

**Review by alyssawilk — APPROVED** [review]
LGTM as long as you'll take a look in the not too distant future.  May be a preexisting bug but I'd rather have an untested fix than a known bug, and I'd rather have a tested fix than either :-)

**Comment by mattklein123** [comment]
> LGTM as long as you'll take a look in the not too distant future. May be a preexisting bug but I'd rather have an untested fix than a known bug, and I'd rather have a tested fix than either :-)

Yeah I will do it. I would like to land this to avoid merge conflicts that keep happening.

**Pull request merged**
